### ISSUE TITLE
EOFError while invoke Docker::Container.create

### ISSUE THREAD

**Issue by plusun** [swipely/docker-api#293]
I wrote a piece of code to create a container from existing image like this:
```
container = Docker::Container.create('Cmd' => ['ls'], 'Image' => 'busybox')
```
However, when I run this line of code, I got the following error messages:
```
/root/.gem/ruby/2.2.0/gems/excon-0.45.4/lib/excon/socket.rb:45:in `read_nonblock': end of file reached (EOFError) (Excon::Errors::SocketError)
        from /root/.gem/ruby/2.2.0/gems/excon-0.45.4/lib/excon/socket.rb:45:in `readline'
        from /root/.gem/ruby/2.2.0/gems/excon-0.45.4/lib/excon/response.rb:53:in `parse'
        from /root/.gem/ruby/2.2.0/gems/excon-0.45.4/lib/excon/middlewares/response_parser.rb:6:in `response_call'
        from /root/.gem/ruby/2.2.0/gems/docker-api-1.22.0/lib/excon/middlewares/hijack.rb:45:in `response_call'
        from /root/.gem/ruby/2.2.0/gems/excon-0.45.4/lib/excon/connection.rb:372:in `response'
        from /root/.gem/ruby/2.2.0/gems/excon-0.45.4/lib/excon/connection.rb:236:in `request'
        from /root/.gem/ruby/2.2.0/gems/docker-api-1.22.0/lib/docker/connection.rb:40:in `request'
        from /root/.gem/ruby/2.2.0/gems/docker-api-1.22.0/lib/docker/connection.rb:65:in `block (2 levels) in <class:Connection>'
        from /root/.gem/ruby/2.2.0/gems/docker-api-1.22.0/lib/docker/container.rb:228:in `create'
        from ./xxx.rb:xx:in `<main>'
```
and the Docker daemon also seems to throw a runtime error:
```
2015/07/17 16:47:22 http: panic serving @: runtime error: invalid memory address or nil pointer dereference
goroutine 78 [running]:
net/http.func·011()
        /usr/local/go/src/net/http/server.go:1130 +0xbb
github.com/docker/docker/api/server.adjustCpuShares(0xc2081628a7, 0x4, 0x0)
        /go/src/github.com/docker/docker/api/server/server_linux.go:112 +0x61
github.com/docker/docker/api/server.(*Server).postContainersCreate(0xc208076cc0, 0xc2081628a7, 0x4, 0x7fde9a485dd8, 0xc208199cc0, 0xc2080b1a00, 0xc208162a20, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/server.go:923 +0x21a
github.com/docker/docker/api/server.*Server.(github.com/docker/docker/api/server.postContainersCreate)·fm(0xc2081628a7, 0x4, 0x7fde9a485dd8, 0xc208199cc0, 0xc2080b1a00, 0xc208162a20, 0x0, 0x0)
        /go/src/github.com/docker/docker/api/server/server.go:1566 +0x7b
github.com/docker/docker/api/server.func·008(0x7fde9a485dd8, 0xc208199cc0, 0xc2080b1a00)
        /go/src/github.com/docker/docker/api/server/server.go:1522 +0xc8f
net/http.HandlerFunc.ServeHTTP(0xc2081c4040, 0x7fde9a485dd8, 0xc208199cc0, 0xc2080b1a00)
        /usr/local/go/src/net/http/server.go:1265 +0x41
github.com/gorilla/mux.(*Router).ServeHTTP(0xc2080a19a0, 0x7fde9a485dd8, 0xc208199cc0, 0xc2080b1a00)
        /go/src/github.com/docker/docker/vendor/src/github.com/gorilla/mux/mux.go:98 +0x297
net/http.serverHandler.ServeHTTP(0xc2081aa060, 0x7fde9a485dd8, 0xc208199cc0, 0xc2080b1a00)
        /usr/local/go/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc208199b80)
        /usr/local/go/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:1751 +0x35e
```

So, I think it seems like some errors on messages between API and server. Could you please  tell me the exact reasons or how to solve it? Thanks a lot!

**Comment by tlunter** [comment]
This looks like the docker deamon is dying and that's why the API is failing.  What version of docker are you running? Does the docker CLI work?

**Comment by plusun** [comment]
Yes, at first glance, I also thought it was dying... However, after this exception, I then ran command lines like `sudo docker images`, and it works well. So it still works well.

The docker's version is 1.7.1, cloned and compiled from the Github source.

**Comment by srust** [comment]
Hi,

This appears to be a combination of a bug in the backend docker daemon (it has a null pointer exception) to the invalid request, coupled with the client (docker-api) not specifying the 'HostConfig' object client side in the 'create' request.

I've created a pull request that fixes it for me and allows docker-api to work with the latest docker development version.

https://github.com/swipely/docker-api/pull/296

**Comment by tlunter** [comment]
Hey guys, let me know if this gets fixed in the release of 1.8.0.  @srust's fork works for now if you're hitting this issue, but I'd prefer this be fixed upstream if possible.

**Comment by blockbridge** [comment]
Thanks @tlunter. Will let you know. I've opened an issue with upstream.

**Comment by srust** [comment]
Thanks @tlunter, will let you know. I've opened an issue with upstream.

**Comment by srust** [comment]
FYI, the fix for this has been merged into docker master, scheduled for 1.8.

https://github.com/docker/docker/pull/14917
